src/mgr.sol

```diff
@@ -186,124 +189,117 @@ contract TinlakeManager {
+        urn.lock(wad);
+        live = true;
+    }
+
     // --- Vault Operation---
     // join & exit move the gem directly into/from the urn
     function join(uint256 wad) public operatorOnly {
         require(safe && live, "TinlakeManager/bad-state");
         require(int256(wad) >= 0, "TinlakeManager/overflow");
         gem.transferFrom(msg.sender, address(this), wad);
-        vat.slip(ilk, address(this), int256(wad));
-        vat.frob(ilk, address(this), address(this), address(this), int256(wad), 0);
         emit Join(wad);
     }
 
     function exit(uint256 wad) public operatorOnly {
         require(safe && live, "TinlakeManager/bad-state");
         require(wad <= 2 ** 255, "TinlakeManager/overflow");
-        vat.frob(ilk, address(this), address(this), address(this), -int256(wad), 0);
-        vat.slip(ilk, address(this), -int256(wad));
         gem.transfer(msg.sender, wad);
         emit Exit(wad);
     }
 
     // draw & wipe call daiJoin.exit/join immediately
     function draw(uint256 wad) public operatorOnly {
         require(safe && live, "TinlakeManager/bad-state");
-        (, uint256 rate, , , ) = vat.ilks(ilk);
-        uint256 dart = divup(mul(RAY, wad), rate);
-        require(int256(dart) >= 0, "TinlakeManager/overflow");
-        vat.frob(ilk, address(this), address(this), address(this), 0, int256(dart));
-        daiJoin.exit(msg.sender, wad);
+        urn.draw(wad);
+        dai.transferFrom(address(this), msg.sender, wad);
         emit Draw(wad);
     }
 
     function wipe(uint256 wad) public operatorOnly {
         require(safe && live, "TinlakeManager/bad-state");
         dai.transferFrom(msg.sender, address(this), wad);
-        daiJoin.join(address(this), wad);
-        (,uint256 rate, , , ) = vat.ilks(ilk);
-        uint256 dart = mul(RAY, wad) / rate;
-        require(dart <= 2 ** 255, "TinlakeManager/overflow");
-        vat.frob(ilk, address(this), address(this), address(this), 0, -int256(dart));
+        urn.wipe(wad);
         emit Wipe(wad);
     }
 
+    function free(uint256 wad)  public operatorOnly {
+        urn.quit();
+    }
+
     // --- Administration ---
     function setOperator(address newOperator) external operatorOnly  {
         operator = newOperator;
         emit SetOperator(newOperator);
     }
 
     function migrate(address dst) public auth  {
-        vat.hope(dst);
         dai.approve(dst, uint256(-1));
         gem.approve(dst, uint256(-1));
         live = false;
         emit Migrate(dst);
     }
 
     function file(bytes32 what, address data) public auth {
-        require(live, "TinlakeManager/not-live");
         emit File(what, data);
-        if (what == "vow") vow = data;
-        else if (what == "daiJoin") daiJoin = JoinLike(data);
+        if (what == "daiJoin") daiJoin = JoinLike(data);
         else if (what == "end")  end = EndLike(data);
+        else if (what == "urn") {
+            urn = MIP21UrnLike(data);
+            dai.approve(data, uint256(-1));
```

Comment:
```suggestion
            dai.approve(data, uint256(-1));
            // rwaUrn allowance for rwaToken
            rwaToken.approve(address(urn), uint(-1));

```

Reply:
Not necessary anymore. We do it in `lock()`